kaldb/src/main/java/com/slack/kaldb/preprocessor/PreprocessorRateLimiter.java

```diff
@@ -0,0 +1,110 @@
+package com.slack.kaldb.preprocessor;
+
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.RateLimiter;
+import com.slack.service.murron.trace.Trace;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.kafka.streams.kstream.Predicate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The PreprocessorRateLimiter provides a thread-safe Kafka streams predicate for determining if a
+ * Kafka message should be filtered or not. Metrics are provided for visibility when a rate limiter
+ * is dropping messages.
+ */
+@ThreadSafe
+@SuppressWarnings("UnstableApiUsage")
+public class PreprocessorRateLimiter {
+  private static final Logger LOG = LoggerFactory.getLogger(PreprocessorRateLimiter.class);
+
+  private final MeterRegistry meterRegistry;
+  private final int preprocessorCount;
+
+  public static final String MESSAGES_DROPPED = "preprocessor_rate_limit_messages_dropped";
+  public static final String BYTES_DROPPED = "preprocessor_rate_limit_bytes_dropped";
+
+  public enum MessageDropReason {
+    NOT_PROVISIONED,
+    OVER_LIMIT
+  }
+
+  public PreprocessorRateLimiter(final MeterRegistry meterRegistry, final int preprocessorCount) {
+    Preconditions.checkArgument(preprocessorCount > 0, "Preprocessor count must be greater than 0");
+
+    this.meterRegistry = meterRegistry;
+    this.preprocessorCount = preprocessorCount;
+  }
+
+  public Predicate<String, Trace.Span> createRateLimiter(
+      Map<String, Long> serviceNameToThroughput) {
+    Map<String, RateLimiter> rateLimiterMap =
```

Comment:
The map is only instantiated once, as this returns a new anonymous class of `Predicate<String, Trace.Span>()` which is then invoked for each run. In the event a service is added, rate limits change, or other mutations are made to the service metadata it will trigger a new invocation of `load()` which rebuilds this map.

This specific use case is also covered in `PreprocessorServiceIntegrationTest.shouldProcessMessageStartToFinish`,  where we first initialize a topic to a very low throughput, increase the throughput, and then assert that the original rate limit was not applied.